common/nym_offline_compact_ecash/src/scheme/setup.rs

```diff
@@ -0,0 +1,560 @@
+/// Otherwise, returns an error describing the nature of the failure.
+///
+/// # Panics
+///
+/// The function may panic if there is an issue with converting bytes to Scalar during initialization.
+pub fn aggregate_indices_signatures(
+    params: &Parameters,
+    vk: &VerificationKeyAuth,
+    signatures: &[(u64, VerificationKeyAuth, Vec<PartialCoinIndexSignature>)],
+) -> Result<Vec<CoinIndexSignature>> {
+    // Check if all indices are unique
+    if signatures
+        .iter()
+        .map(|(index, _, _)| index)
+        .unique()
+        .count()
+        != signatures.len()
+    {
+        return Err(CompactEcashError::CoinIndices(
+            "Not enough unique indices shares".to_string(),
+        ));
+    }
+
+    // Evaluate at 0 the Lagrange basis polynomials k_i
+    let coefficients = generate_lagrangian_coefficients_at_origin(
+        &signatures
+            .iter()
+            .map(|(index, _, _)| *index)
+            .collect::<Vec<_>>(),
+    );
+
+    // Verify that all signatures are valid
+    signatures
+        .par_iter()
+        .try_for_each(|(_, vk_auth, partial_signatures)| {
+            verify_coin_indices_signatures(params, vk, vk_auth, partial_signatures)
+        })?;
+
+    // Pre-allocate vectors
+    let mut aggregated_coin_signatures: Vec<CoinIndexSignature> =
+        Vec::with_capacity(params.get_total_coins() as usize);
+
+    for l in 0..params.get_total_coins() {
+        // Compute the hash h
+        let mut concatenated_bytes =
+            Vec::with_capacity(vk.to_bytes().len() + l.to_le_bytes().len());
+        concatenated_bytes.extend_from_slice(&vk.to_bytes());
+        concatenated_bytes.extend_from_slice(&l.to_le_bytes());
+        let h = hash_g1(concatenated_bytes);
+
+        // Collect the partial signatures for the same coin index
+        let collected_at_l: Vec<_> = signatures
+            .iter()
+            .filter_map(|(_, _, inner_vec)| inner_vec.get(l as usize))
+            .cloned()
+            .collect();
+
+        // Aggregate partial signatures for each coin index
+        let aggr_s: G1Projective = coefficients
+            .iter()
+            .zip(collected_at_l.iter())
+            .map(|(coeff, sig)| sig.s * coeff)
+            .sum();
+        let aggr_sig = CoinIndexSignature { h, s: aggr_s };
+        aggregated_coin_signatures.push(aggr_sig);
+    }
+    verify_coin_indices_signatures(params, vk, vk, &aggregated_coin_signatures)?;
+    Ok(aggregated_coin_signatures)
+}
+
+/// Generates parameters for the scheme setup.
+///
+/// # Arguments
+///
+/// * `total_coins` - it is the number of coins in a freshly generated wallet. It is the public parameter of the scheme.
+///
+/// # Returns
+///
+/// A `Parameters` struct containing group parameters, public key, the number of signatures (`total_coins`),
+/// and a map of signatures for each index `l`.
+///
+pub fn setup(total_coins: u64) -> Parameters {
```

Comment:
In theory, I think yes. In practice you could not spend anything so it's a bit useless. An `assert` sounds good, to avoid propagating changes to a gazilion places. Fixed in 8d39401cc0208e93675cf0bff508f99f59f04a0d